Simple c++ base class for processing algorithms

nyalldawson opened the pull request:
This PR introduces a new c++ base class for processing algorithms. It's another step along the way in porting the processing framework to c++.

So far it's quite basic - it just handles groups/names/flags/icons. There's no logic here yet. But I need this part merged and a c++ algorithm class in order to complete the port of the providers and alglist to c++.

nyalldawson committed:
Simple c++ base class for processing algorithms

nyalldawson committed:
Move tag handling to QgsProcessingAlgorithm

nyalldawson committed:
Move algorithm icon handling to QgsProcessingAlgorithm

nyalldawson committed:
Move algorithm group handling to QgsProcessingAlgorithm

nyalldawson committed:
Port handling of algorithm names to QgsProcessingAlgorithm

nyalldawson committed:
Move flags (hide from toolbox, etc) to QgsProcessingAlgorithm

m-kuhn commented on src/core/processing/qgsprocessingalgorithm.cpp:
```diff
@@ -0,0 +1,34 @@
+  return QgsApplication::getThemeIcon( "/processingAlgorithm.svg" );
+}
+
+QString QgsProcessingAlgorithm::svgIconPath() const
```
Any reason to explicitly insist on svg here?

nyalldawson replied:
It's a request from @nirvn, used for high resolution exports of the model

m-kuhn commented on src/core/processing/qgsprocessingalgorithm.h:
```diff
@@ -0,0 +1,105 @@
+      FlagHideFromToolbox = 1 << 1, //!< Algorithm should be hidden from the toolbox
+      FlagHideFromModeler = 1 << 2, //!< Algorithm should be hidden from the modeler
+      FlagSupportsBatch = 1 << 3,  //!< Algorithm supports batch mode
+      FlagDeprecated = FlagHideFromToolbox | FlagHideFromModeler, //!< Algorithm is deprecated
```
Nice idea to put it this way.

m-kuhn commented:
Yes please

nyalldawson committed:
Use formatted name for algorithm names

Avoids conversion of names later and gives clearer separation
between visible names and internal names

nyalldawson committed:
Sipify QgsProcessingAlgorithm

nyalldawson committed:
Move format handling flags to QgsAlgorithmProvider

nyalldawson committed:
Fix failing test

nyalldawson committed:
Update PreconfiguredAlgorithm for new API

nyalldawson commented:
@volaya @alexbruy any thoughts? I'd like to merge this as more followup changes are inbound shortly :)

alexbruy commented:
Looks good for me
